<span style="font-family: monospace">

## Business Problem

<span style="color: #77BBFF; font-size: 12px">
Analyze the data and generate insights that could help Netflix in deciding which type of shows/movies to produce and how they can grow the business in different countries.
</span>

## Hints

<span style="color: #77BBFF; font-size: 12px">
1. The exploration should have a goal. As you explore the data, keep in mind that you want to answer which type of shows to produce and how to grow the business.
<br>
2. Ensure each recommendation is backed by data. The company is looking for data-driven insights, not personal opinions or anecdotes.
<br>
3. Assume that you are presenting your findings to business executives who have only a basic understanding of data science. Avoid unnecessary technical jargon.
<br>
4. Start by exploring a few questions: What type of content is available in different countries?
<br>
<span style="padding-left: 3em">
    a. How has the number of movies released per year changed over the last 20-30 years?
</span>
<br>
<span style="padding-left: 3em">
    b. Comparison of tv shows vs. movies.
</span>
<br>
<span style="padding-left: 3em">
    c. What is the best time to launch a TV show?
</span>
<br>
<span style="padding-left: 3em">
    d. Analysis of actors/directors of different types of shows/movies.
</span>
<br>
<span style="padding-left: 3em">
    e. Does Netflix has more focus on TV Shows than movies in recent years.
</span>
<br>
<span style="padding-left: 3em">
    f. Understanding what content is available in different countries.
</span>
</span>

## Column Description

<span style="color: #77BBFF; font-size: 12px">
1. show_id      : Unique ID for every Movie / TV Show
<br>
2. type         : Identifier - A Movie or TV Show
<br>
3. title        : Title of the Movie / Tv Show
<br>
4. director     : Director of the Movie
<br>
5. cast         : Actors involved in the movie/show
<br>
6. country      : Country where the movie/show was produced
<br>
7. date_added   : Date it was added on Netflix
<br>
8. release_year : Actual Release year of the movie/show
<br>
9. rating       : TV Rating of the movie/show
<br>
10. duration     : Total Duration - in minutes or number of seasons
<br>
11. listed_in    : Genre
<br>
12. description  : The summary description
</span>

## Evaluation Criteria (100 Points):

<span style="color: #77BBFF; font-size: 12px">
1. Defining Problem Statement and Analysing basic metrics. (10 Points)
<br>
2. Observations on the shape of data, data types of all the attributes, conversion of categorical attributes to 'category' (If required), missing value detection, statistical summary. (10 Points)
<br>
3. Non-Graphical Analysis: Value counts and unique attributes. ​​(10 Points)
<br>
4. Visual Analysis - Univariate, Bivariate after pre-processing of the data. Note: Pre-processing involves unnesting of the data in columns like Actor, Director, Country.
<br>
<span style="padding-left: 3em">
    a. For continuous variable(s): Distplot, countplot, histogram for univariate analysis. (10 Points)
</span>
<br>
<span style="padding-left: 3em">
    b. For categorical variable(s): Boxplot. (10 Points)
</span>
<br>
<span style="padding-left: 3em">
    c. For correlation: Heatmaps, Pairplots. (10 Points)
</span>
<br>
5. Missing Value & Outlier check (Treatment optional). (10 Points)
<br>
6. Insights based on Non-Graphical and Visual Analysis. (10 Points)
<br>
<span style="padding-left: 3em">
    a. Comments on the range of attributes.
</span>
<br>
<span style="padding-left: 3em">
    b. Comments on the distribution of the variables and relationship between them.
</span>
<br>
<span style="padding-left: 3em">
    c. Comments for each univariate and bivariate plot.
</span>
<br>
7. Business Insights (10 Points) - Should include patterns observed in the data along with what you can infer from it.
<br>
8. Recommendations (10 Points) - Actionable items for business. No technical jargon. No complications. Simple action items that everyone can understand.
</span>

## Submission Process:

<span style="color: #77BBFF; font-size: 12px">
1. Type your insights and recommendations in the rich-text editor.
<br>
2. Convert your jupyter notebook into PDF (Save as PDF using Chrome browser’s Print command), upload it in your Google Drive (set the permission to allow public access), and paste that link in the text editor.
<br>
3. Alternatively, you can directly submit your PDF on the portal.
<br>
4. Optionally, you may add images/graphs in the text editor by taking screenshots or saving matplotlib graphs using plt.savefig(...).
<br>
5. After submitting, you will not be allowed to edit your submission.
</span>

</span>

<span style="font-family: monospace; color: #11DDCC; font-size: 40px">
Let's Start!
</span>

In [1]:
# importing the required libraries
import re, numpy as np, pandas as pd, plotly.express as px, plotly.graph_objects as go

# reading the data from the `data.csv` saved in the same folder as this jupyter-notebook
df = pd.read_csv('data.csv')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
The First Step towards solving any Business Problem through Data is Exploratory Data Analysis (EDA) and the First Step towards EDA is basis analysis of the data (number of records, number of features and their corresponding data types), locating and eliminating Missing Values, and transforming features into something which is explorable and meaningfull.
<br>
<br>
We will now work on getting answers of all the points we just mentioned and these answers will help us give answer to Q2 and Q5 of the Evaluation Criteria.
<br>
<br>
Q2: Observations on the shape of data, data types of all the attributes, conversion of categorical attributes to 'category' (If required), missing value detection, statistical summary. (10 Points)
<br>
Q5: Missing Value & Outlier check (Treatment optional). (10 Points)
</span>

In [2]:
# printing sample data
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [3]:
# printing the shape of the data
df.shape

(8807, 12)

In [4]:
# checking columns' datatype
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

<span style="font-family: monospace; color: #11DDCC; font-size: 40px">
Data Cleaning and Preprocessing
</span>

In [5]:
# dropping `show_id` column because it won't help us in our analysis as it is just an unique identifier
df.drop('show_id', axis=1, inplace=True)

# converting `release_year` to `object` because we will be using it as a categorical feature
df['release_year'] = df['release_year'].astype(object)

# converting `date_added` to `datetime` because we will be performing yearly/monthly analysis
df['date_added'] = pd.to_datetime(df['date_added'], infer_datetime_format=True).dt.date

In [6]:
# since now we have all our columns as categorical, `df.describe` would print statistical summary of the data without explicitly mentioning `include=object`
df.describe(datetime_is_numeric=False)

type                 title       director                cast  \
count    8807                  8807           6173                7982   
unique      2                  8807           4528                7692   
top     Movie  Dick Johnson Is Dead  Rajiv Chilaka  David Attenborough   
freq     6131                     1             19                  19   

              country  date_added  release_year rating  duration  \
count            7976        8797          8807   8803      8804   
unique            748        1714            74     17       220   
top     United States  2020-01-01          2018  TV-MA  1 Season   
freq             2818         110          1147   3207      1793   

                           listed_in  \
count                           8807   
unique                           514   
top     Dramas, International Movies   
freq                             362   

                                              description  
count                                                8807  
unique                                               8775  
top     Paranormal activity at a lush, abandoned prope...  
freq                                                    4

In [7]:
# checking for missing values
df.isna().sum()

type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [8]:
# working on `duration` column and checking NaNs values
df[df['duration'].isna()]

type                                 title    director        cast  \
5541  Movie                       Louis C.K. 2017  Louis C.K.  Louis C.K.   
5794  Movie                 Louis C.K.: Hilarious  Louis C.K.  Louis C.K.   
5813  Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.  Louis C.K.   

            country  date_added release_year  rating duration listed_in  \
5541  United States  2017-04-04         2017  74 min      NaN    Movies   
5794  United States  2016-09-16         2010  84 min      NaN    Movies   
5813  United States  2016-08-15         2015  66 min      NaN    Movies   

                                            description  
5541  Louis C.K. muses on religion, eternal love, gi...  
5794  Emmy-winning comedy writer Louis C.K. brings h...  
5813  The comic puts his trademark hilarious/thought...

In [9]:
# we saw that `duration` is filled in `rating` column. hence correcting the mistake
for idx in df[df['duration'].isna()].index:
    df.loc[idx, 'duration'] = df.loc[idx, 'rating']
    df.loc[idx, 'rating'] = np.nan

# again checking for NaNs values and this time it should be an empty dataframe and hence no error should be raised output
assert df[df['duration'].isna()].shape[0] == 0

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Transforming <span style="color: #ECE93D"> duration </span> Column
</span>

In [10]:
# we will first check if all the corresponding to `type=Movie` and `type=TV Show` (because these are the only two categories for `type` column as can be verified
# from `df['type'].unique()`) ends in `min` and `Season` or `Seasons` respectively

# filtering `Movie` and checking some conditions to successfully remove 'min' from `duration`
ser = df[df['type'] == 'Movie']['duration']
ser = ser[ser.notna()].str.split()
assert set(ser.apply(lambda x: len(x)).to_list()) == {2,}
assert set(ser.apply(lambda x: x[1]).to_list()) == {'min',}

# filtering `TV Show` and checking some conditions to successfully remove 'Season' or 'Seasons' from `duration`
ser = df[df['type'] == 'TV Show']['duration']
ser = ser[ser.notna()].str.split()
assert set(ser.apply(lambda x: len(x)).to_list()) == {2,}
assert set(ser.apply(lambda x: x[1]).to_list()) == {'Season', 'Seasons', }

# stripping off 'min', 'Season', and 'Seasons' from `duration` column
df['duration'] = df['duration'].str.split().apply(lambda x: x[0]).astype(int)

df.head()

type                  title         director  \
0    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1  TV Show          Blood & Water              NaN   
2  TV Show              Ganglands  Julien Leclercq   
3  TV Show  Jailbirds New Orleans              NaN   
4  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

   date_added release_year rating  duration  \
0  2021-09-25         2020  PG-13        90   
1  2021-09-24         2021  TV-MA         2   
2  2021-09-24         2021  TV-MA         1   
3  2021-09-24         2021  TV-MA         1   
4  2021-09-24         2021  TV-MA         2   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Transforming dataframe basis the  <span style="color: #ECE93D"> description </span> Column
</span>

In [11]:
# in `df.describe` we saw that in the column `description` top entry occurred 4 times. let's investigate a bit on more this.
print(df[df['description'].duplicated()].shape)
df[df['description'].duplicated()].head()

(32, 11)


type                    title                  director  \
79   Movie  Tughlaq Durbar (Telugu)  Delhiprasad Deenadayalan   
237  Movie          Boomika (Hindi)       Rathindran R Prasad   
238  Movie      Boomika (Malayalam)       Rathindran R Prasad   
239  Movie         Boomika (Telugu)       Rathindran R Prasad   
851  Movie         99 Songs (Tamil)                       NaN   

                                                  cast country  date_added  \
79          Vijay Sethupathi, Parthiban, Raashi Khanna     NaN  2021-09-11   
237  Aishwarya Rajesh, Vidhu, Surya Ganapathy, Madh...     NaN  2021-08-23   
238  Aishwarya Rajesh, Vidhu, Surya Ganapathy, Madh...     NaN  2021-08-23   
239  Aishwarya Rajesh, Vidhu, Surya Ganapathy, Madh...     NaN  2021-08-23   
851                                                NaN     NaN  2021-05-21   

    release_year rating  duration  \
79          2021  TV-14       145   
237         2021  TV-14       122   
238         2021  TV-14       122   
239         2021  TV-14       122   
851         2021  TV-14       131   

                                          listed_in  \
79           Comedies, Dramas, International Movies   
237  Horror Movies, International Movies, Thrillers   
238  Horror Movies, International Movies, Thrillers   
239  Horror Movies, International Movies, Thrillers   
851  Dramas, International Movies, Music & Musicals   

                                           description  
79   A budding politician has devious plans to rise...  
237  Paranormal activity at a lush, abandoned prope...  
238  Paranormal activity at a lush, abandoned prope...  
239  Paranormal activity at a lush, abandoned prope...  
851  Challenged to compose 100 songs before he can ...

In [12]:
# on perusing the above dataframe, one can observe that these are rows all corresponds to a particular set of movies released in different languages.
# also, for the same movie (based on description) there are NaNs present in different rows. what we need to do is examine these rows one by one and fill NaNs.

all_duplicated_description = df[df['description'].duplicated()]['description'].to_list()
all_duplicated_description = {desc: all_duplicated_description.count(desc) for desc in all_duplicated_description}

replace_desc_dict, custom_replace_desc_dict = dict(), dict()

for idx, desc in enumerate(all_duplicated_description.keys()):
    temp_df = df[df['description'] == desc]
    
    # checking if for same `description` we have same `type`, `director`, and `country` or not
    # if all these are same or missing, we are treating different title as similar to one another and updating them in `replace_desc_dict`
    # otherwise treating them as different and updating them in `custom_replace_desc_dict`
    if (temp_df['type'].nunique() < 2) and (temp_df['director'].nunique() < 2) and (temp_df['country'].nunique() < 2):
        replace_desc_dict[desc] = all_duplicated_description[desc]
    else:
        custom_replace_desc_dict[desc] = all_duplicated_description[desc]
len(replace_desc_dict), len(custom_replace_desc_dict)

(25, 2)

In [13]:
# forming a common strategy to fill NaNs in rows corresponding to 25 descriptions in `replace_desc_dict` with `mode` of each column where not all values are NaNs.
for desc in list(replace_desc_dict.keys()):
    temp_df = df[df['description'] == desc]
    for column_name in df.columns:
        col_filler = df.loc[temp_df.index][column_name]
        col_filler_value = col_filler[col_filler.notna()].mode()
        if col_filler_value.to_list() != []:
            df.loc[temp_df.index, column_name] = col_filler_value[0]

In [14]:
# analyzing rows corresponding to first `description` in `custom_replace_desc_dict`
desc = list(custom_replace_desc_dict.keys())[0]

temp_df = df[df['description'] == desc]
# we saw that there are a total of 3 rows corresponding to this `description` and 2 are of `type=Movie` and 1 is of `type=TV Show`.
# hence filtering only for `type=Movie` and applying the same strategy as that applied above

temp_df = temp_df[temp_df['type'] == 'Movie']
for column_name in df.columns:
    col_filler = df.loc[temp_df.index][column_name]
    col_filler_value = col_filler[col_filler.notna()].mode()
    if col_filler_value.to_list() != []:
        df.loc[temp_df.index, column_name] = col_filler_value[0]

In [15]:
# analyzing rows corresponding to second `description` in `custom_replace_desc_dict`
desc = list(custom_replace_desc_dict.keys())[1]
temp_df = df[df['description'] == desc]
temp_df

# nothing we can do here because there are a total of 2 rows corresponding to this `description` with 1 each belonging to `type=Movie` and `type=TV Show`.
# hence leaving as it is.

type              title       director  \
2336    Movie  Thackeray (Hindi)  Abhijit Panse   
8173  TV Show          Thackeray            NaN   

                                                   cast country  date_added  \
2336  Nawazuddin Siddiqui, Amrita Rao, Rajeev Panday...   India  2020-06-26   
8173                                                NaN   India  2019-05-25   

     release_year rating  duration                     listed_in  \
2336         2019  TV-14       135  Dramas, International Movies   
8173         2019  TV-MA         1                      TV Shows   

                                            description  
2336  From controversial cartoonist to powerful Mumb...  
8173  From controversial cartoonist to powerful Mumb...

In [16]:
# checking shape of dataframe before and after dropping duplicates
print(df.shape, ' - Shape of DataFrame with duplicates.')
df.drop_duplicates(inplace=True)
print(df.shape, ' - Shape of DataFrame without duplicates.')

(8807, 11)  - Shape of DataFrame with duplicates.
(8777, 11)  - Shape of DataFrame without duplicates.


In [17]:
# checking `df.describe()` for further steps
df.describe(include=object)

type                 title       director                cast  \
count    8777                  8777           6150                7956   
unique      2                  8777           4528                7687   
top     Movie  Dick Johnson Is Dead  Rajiv Chilaka  David Attenborough   
freq     6105                     1             19                  19   

              country  date_added  release_year rating  \
count            7963        8767          8777   8770   
unique            748        1714            74     14   
top     United States  2020-01-01          2018  TV-MA   
freq             2816         110          1140   3202   

                           listed_in  \
count                           8777   
unique                           514   
top     Dramas, International Movies   
freq                             362   

                                              description  
count                                                8777  
unique                                               8775  
top     From controversial cartoonist to powerful Mumb...  
freq                                                    2

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Filling <span style="color: #ECE93D">NaNs</span> in <span style="color: #ECE93D">director</span>, <span style="color: #ECE93D">cast</span>, <span style="color: #ECE93D">country</span> and <span style="color: #ECE93D">rating</span> Columns
</span>

In [18]:
# let's fill NaNs in `director` and `cast` as 'Anonymous' and in `country` as `Not Available`
df['director'].fillna('Anonymous', inplace=True)
df['cast'].fillna('Anonymous', inplace=True)
df['country'].fillna('Not Available', inplace=True)
df['rating'].fillna('Not Available', inplace=True)
# now only 2 more columns contain NaNs and these are `date_added` and `rating`

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Filling <span style="color: #ECE93D"> NaTs </span> in the <span style="color: #ECE93D"> date_added </span> Column
</span>

In [19]:
df[df['date_added'].isna()]

type                                        title   director  \
6066  TV Show  A Young Doctor's Notebook and Other Stories  Anonymous   
6174  TV Show              Anthony Bourdain: Parts Unknown  Anonymous   
6795  TV Show                                      Frasier  Anonymous   
6806  TV Show                                      Friends  Anonymous   
6901  TV Show                              Gunslinger Girl  Anonymous   
7196  TV Show                                     Kikoriki  Anonymous   
7254  TV Show                          La Familia P. Luche  Anonymous   
7406  TV Show                                        Maron  Anonymous   
7847  TV Show                                 Red vs. Blue  Anonymous   
8182  TV Show                 The Adventures of Figaro Pho  Anonymous   

                                                   cast         country  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...  United Kingdom   
6174                                   Anthony Bourdain   United States   
6795  Kelsey Grammer, Jane Leeves, David Hyde Pierce...   United States   
6806  Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   United States   
6901  Yuuka Nanri, Kanako Mitsuhashi, Eri Sendai, Am...           Japan   
7196                                      Igor Dmitriev   Not Available   
7254  Eugenio Derbez, Consuelo Duval, Luis Manuel Áv...   United States   
7406  Marc Maron, Judd Hirsch, Josh Brener, Nora Zeh...   United States   
7847  Burnie Burns, Jason Saldaña, Gustavo Sorola, G...   United States   
8182  Luke Jurevicius, Craig Behenna, Charlotte Haml...       Australia   

     date_added release_year rating  duration  \
6066        NaT         2013  TV-MA         2   
6174        NaT         2018  TV-PG         5   
6795        NaT         2003  TV-PG        11   
6806        NaT         2003  TV-14        10   
6901        NaT         2008  TV-14         2   
7196        NaT         2010   TV-Y         2   
7254        NaT         2012  TV-14         3   
7406        NaT         2016  TV-MA         4   
7847        NaT         2015     NR        13   
8182        NaT         2015  TV-Y7         2   

                                              listed_in  \
6066           British TV Shows, TV Comedies, TV Dramas   
6174                                         Docuseries   
6795                     Classic & Cult TV, TV Comedies   
6806                     Classic & Cult TV, TV Comedies   
6901                       Anime Series, Crime TV Shows   
7196                                           Kids' TV   
7254  International TV Shows, Spanish-Language TV Sh...   
7406                                        TV Comedies   
7847  TV Action & Adventure, TV Comedies, TV Sci-Fi ...   
8182                              Kids' TV, TV Comedies   

                                            description  
6066  Set during the Russian Revolution, this comic ...  
6174  This CNN original series has chef Anthony Bour...  
6795  Frasier Crane is a snooty but lovable Seattle ...  
6806  This hit sitcom follows the merry misadventure...  
6901  On the surface, the Social Welfare Agency appe...  
7196  A wacky rabbit and his gang of animal pals hav...  
7254  This irreverent sitcom featues Ludovico, Feder...  
7406  Marc Maron stars as Marc Maron, who interviews...  
7847  This parody of first-person shooter games, mil...  
8182  Imagine your worst fears, then multiply them: ...

In [20]:
# what we have in above dataframe is that for all the records missing `date_added` entry, we have information of `release_year`.
# what we can do here is that fill NaNs with release_year-12-31, i.e., 31st December of the release year

indices_to_modify = df[df['date_added'].isna()].index

df.loc[indices_to_modify, 'date_added'] = df.loc[indices_to_modify]['release_year'].apply(lambda x: str(x)+'-12-31')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Transforming <span style="color: #ECE93D">director</span>, <span style="color: #ECE93D">cast</span>, <span style="color: #ECE93D">listed_in </span>, and <span style="color: #ECE93D">country</span> Columns
</span>

In [21]:
for column_name in ['director', 'cast', 'listed_in', 'country']:
    df[column_name] = df[column_name].apply(lambda x: x.split(','))

df = df.explode('director').explode('cast').explode('listed_in').explode('country')

for col in df.select_dtypes(object):
    df[col] = df[col].astype(str).str.strip()

df.drop_duplicates(inplace=True)
df.reset_index(drop=True, inplace=True)
df.shape

(201316, 11)

In [22]:
df.head()

type                 title         director         cast        country  \
0    Movie  Dick Johnson Is Dead  Kirsten Johnson    Anonymous  United States   
1  TV Show         Blood & Water        Anonymous   Ama Qamata   South Africa   
2  TV Show         Blood & Water        Anonymous   Ama Qamata   South Africa   
3  TV Show         Blood & Water        Anonymous   Ama Qamata   South Africa   
4  TV Show         Blood & Water        Anonymous  Khosi Ngema   South Africa   

   date_added release_year rating  duration               listed_in  \
0  2021-09-25         2020  PG-13        90           Documentaries   
1  2021-09-24         2021  TV-MA         2  International TV Shows   
2  2021-09-24         2021  TV-MA         2               TV Dramas   
3  2021-09-24         2021  TV-MA         2            TV Mysteries   
4  2021-09-24         2021  TV-MA         2  International TV Shows   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  After crossing paths at a party, a Cape Town t...  
3  After crossing paths at a party, a Cape Town t...  
4  After crossing paths at a party, a Cape Town t...

In [23]:
df.describe(include=object)

type                        title   director       cast  \
count   201316                       201316     201316     201316   
unique       2                         8777       4994      36433   
top      Movie  Kahlil Gibran's The Prophet  Anonymous  Anonymous   
freq    145305                          700      50497       2139   

              country  date_added release_year  rating listed_in  \
count          201316      201316       201316  201316    201316   
unique            124        1719           74      15        42   
top     United States  2020-01-01         2018   TV-MA    Dramas   
freq            59214        3748        24198   73675     29713   

                                              description  
count                                              201316  
unique                                               8775  
top     A troubled young girl and her mother find sola...  
freq                                                  700

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Okay, so now we are ready for some EDA because we have cleansed the data as much as it was possible. The next step would be observing interesting patterns within our data and perform Univariate, Bivariate, and/or Multivariate analysis.
<br>
<br>
Let's start exploring how are the categories distributed within our <span style="font-family: monospace; color: #ECE93D; font-size: 12px">object</span> type columns. All the columns makes sense to explore category-balance using <span style="font-family: monospace; color: #ECE93D; font-size: 12px">value_counts()</span> method except <span style="font-family: monospace; color: #ECE93D; font-size: 12px">description</span> column because it would be similar to the <span style="font-family: monospace; color: #ECE93D; font-size: 12px">title</span> column.
</span>

<span style="font-family: monospace; color: #11DDCC; font-size: 40px">
Data Analysis
</span>

In [24]:
df['type'].value_counts()

Movie      145305
TV Show     56011
Name: type, dtype: int64

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">Inference A: There are almost 2.5 times Movie titles in our dataset than TV shows. Audience tend to like movies over tv shows.</span>

In [25]:
df['title'].value_counts()

Kahlil Gibran's The Prophet                          700
Holidays                                             504
Movie 43                                             468
The Eddy                                             416
Narcos                                               378
                                                    ... 
Marc Maron: Thinky Pain                                1
Lo and Behold: Reveries of the Connected World         1
Miniforce: Super Dino Power                            1
Edmilson Filho: Notas, Comedy about Relationships      1
Dick Johnson Is Dead                                   1
Name: title, Length: 8777, dtype: int64

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference B: On exploding our data, a row in the original dataframe expanded to as high as 700 rows for title Kahlil Gibran's The Prophet while few rows didn't experienced any change.
</span>

In [26]:
df['director'].value_counts()

Anonymous                  50497
Martin Scorsese              419
Youssef Chahine              409
Cathy Garcia-Molina          356
Steven Spielberg             355
                           ...  
Charlie Siskel                 1
Jonathan Ignatius Green        1
Brendon Marotta                1
Sharon Grimberg                1
Kirsten Johnson                1
Name: director, Length: 4994, dtype: int64

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference C: Martin Scorsese occurs the most - 419 - number of times in our dataset; closely followed by Youssef Chahine - 409 times. These directors have worked in movies or tc shows that star many cast members and belong to many genres.
</span>

In [27]:
df['cast'].value_counts()

Anonymous         2139
Liam Neeson        161
Alfred Molina      160
John Krasinski     139
Salma Hayek        130
                  ... 
Elise Loehnen        1
Wil Willis           1
Samantha Bee         1
Eleanor Rocha        1
Zbyněk Vičar         1
Name: cast, Length: 36433, dtype: int64

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference D: Liam Neeson occurs the most - 161 - number of times in our dataset; closely followed by Alfred Molina - 160 times. These actors have worked in movies or tv shows that are directed by many directors and have many cast members and belong to many genres.
</span>

In [28]:
df['country'].value_counts()

United States     59214
India             22733
United Kingdom    12965
Not Available     11620
Japan              8595
                  ...  
Panama                2
Mongolia              2
Kazakhstan            1
Nicaragua             1
Uganda                1
Name: country, Length: 124, dtype: int64

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference E: Highest number of entries - 59214 - in our dataset corresponds to the US followed by India at 22733. These countries have multi-director, multi-starrer, multi-genre content.
</span>

In [29]:
df['date_added'].value_counts().head(50)

2020-01-01    3748
2019-11-01    2258
2021-07-01    2219
2017-10-01    1899
2021-09-01    1756
2018-03-01    1752
2019-12-31    1695
2019-10-01    1563
2018-10-01    1419
2021-06-02    1260
2021-08-01    1248
2021-01-01    1216
2017-09-01    1210
2018-01-01    1141
2018-11-01    1128
2021-07-06    1061
2019-01-01    1059
2017-08-01    1008
2020-04-01    1000
2019-09-01     970
2020-11-01     896
2017-07-01     888
2020-06-01     879
2020-10-01     867
2019-07-01     866
2018-04-01     847
2021-05-01     825
2017-05-01     796
2019-03-01     792
2021-04-01     792
2020-07-05     790
2019-02-01     787
2019-11-20     761
2020-09-01     753
2018-08-01     753
2018-07-01     747
2016-01-01     726
2018-12-01     715
2017-12-01     713
2021-08-27     702
2019-08-01     696
2021-06-19     685
2019-12-01     680
2017-11-01     664
2019-12-15     640
2020-10-19     628
2020-12-01     615
2021-04-16     610
2017-06-01     597
2017-03-10     593
Name: date_added, dtype: int64

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference F: Most of the movies are added on the platform within first week of a month.
</span>

In [30]:
df['release_year'].value_counts()

2018    24198
2019    21756
2017    20516
2020    19652
2016    18465
        ...  
1947        8
1946        6
1942        6
1943        5
1925        1
Name: release_year, Length: 74, dtype: int64

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference G: Audience prefer newer content.
</span>

In [31]:
df['rating'].value_counts()

TV-MA            73675
TV-14            43597
R                25817
PG-13            16222
TV-PG            14893
PG               10903
TV-Y7             6304
TV-Y              3662
TV-G              2749
NR                1573
G                 1530
NC-17              149
TV-Y7-FV            86
UR                  86
Not Available       70
Name: rating, dtype: int64

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference H: Highest number of entries - 73695 - in our dataset corresponds to the TV-MA followed by TV-14 at 43597. These are two most preferred rating types among audience.
</span>

In [32]:
df['listed_in'].value_counts()

Dramas                          29713
International Movies            28084
Comedies                        20758
International TV Shows          12797
Action & Adventure              12167
Independent Movies               9810
Children & Family Movies         9755
TV Dramas                        8894
Thrillers                        7036
Romantic Movies                  6411
TV Comedies                      4963
Crime TV Shows                   4733
Kids' TV                         4565
Horror Movies                    4538
Sci-Fi & Fantasy                 4029
Romantic TV Shows                3049
Music & Musicals                 3028
Documentaries                    2409
Anime Series                     2313
TV Action & Adventure            2288
Spanish-Language TV Shows        2088
British TV Shows                 1808
Sports Movies                    1531
Classic Movies                   1443
TV Mysteries                     1281
Korean TV Shows                  1122
Cult Movies 

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference I: Highest number of entries - 29713 - in our dataset corresponds to the Dramas followed by Comedy at 20758.
</span>

<span style="font-family: monospace; color: #11DDCC; font-size: 40px">
Data Vizualization
</span>

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Let's start building some intuitive plots to get to know the data better.
</span>

In [33]:
df.head()

type                 title         director         cast        country  \
0    Movie  Dick Johnson Is Dead  Kirsten Johnson    Anonymous  United States   
1  TV Show         Blood & Water        Anonymous   Ama Qamata   South Africa   
2  TV Show         Blood & Water        Anonymous   Ama Qamata   South Africa   
3  TV Show         Blood & Water        Anonymous   Ama Qamata   South Africa   
4  TV Show         Blood & Water        Anonymous  Khosi Ngema   South Africa   

   date_added release_year rating  duration               listed_in  \
0  2021-09-25         2020  PG-13        90           Documentaries   
1  2021-09-24         2021  TV-MA         2  International TV Shows   
2  2021-09-24         2021  TV-MA         2               TV Dramas   
3  2021-09-24         2021  TV-MA         2            TV Mysteries   
4  2021-09-24         2021  TV-MA         2  International TV Shows   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  After crossing paths at a party, a Cape Town t...  
3  After crossing paths at a party, a Cape Town t...  
4  After crossing paths at a party, a Cape Town t...

In [34]:
# defining some functions for plotting graphs with minimal code repetition
def update_fig_layout_wo_legend(fig, title, x_title, y_title):
    fig.update_layout(
        xaxis={'visible': True, 'showticklabels': True},
        font_family='monospace',
        font_color='black',
        title_x=0.50,
        title_y=0.95,
        title_text=title,
        xaxis_title=x_title,
        yaxis_title=y_title,
        plot_bgcolor='rgba(255,255,255,255)',
        paper_bgcolor='rgba(255,255,255,255)',
        showlegend=False,
        # hovermode=False
        )
    fig.show()
    return None

def update_fig_layout_w_legends(fig, title, x_title, y_title):
    fig.update_layout(
        xaxis={'visible': True, 'showticklabels': True},
        font_family='monospace',
        font_color='black',
        title_x=0.50,
        title_y=0.95,
        title_text=title,
        xaxis_title=x_title,
        yaxis_title=y_title,
        plot_bgcolor='rgba(255,255,255,255)',
        paper_bgcolor='rgba(255,255,255,255)',
        showlegend=True,
        # hovermode=False
        )
    fig.show()
    return None

def plot_bar(x, y, title, x_title, y_title):
    fig = px.bar(x=x, y=y, width=1250)
    update_fig_layout_wo_legend(fig, title, x_title, y_title)
    return None

def plot_line(x, y, title, x_title, y_title):
    fig = px.line(x=x, y=y, width=1250)
    update_fig_layout_wo_legend(fig, title, x_title, y_title)
    return None

def plot_histogram(x, title, x_title, y_title):
    fig = px.histogram(x=x)
    update_fig_layout_wo_legend(fig, title, x_title, y_title)
    return None

def plot_box(x, y, title, x_title, y_title):
    fig = px.box(x=x, y=y)
    update_fig_layout_wo_legend(fig, title, x_title, y_title)
    return None

def plot_box_one_variable(x, title, y_title):
    fig = px.box(x)
    update_fig_layout_wo_legend(fig, title, '', y_title)
    return None

def plot_dodge_bar(data, x, y, color, title, x_title, y_title):
    fig = px.bar(data_frame=data,x=x, y=y, color=color, barmode='group')
    update_fig_layout_w_legends(fig, title, x_title, y_title)
    return None

def plot_bar_with_running_total(x, y, title, x_title, y_title):
    fig = go.Figure()
    fig.add_trace(go.Scatter(x=x, y=y.cumsum()))
    fig.add_trace(go.Bar(x=x, y=y))
    update_fig_layout_wo_legend(fig, title, x_title, y_title)
    return None

def plot_heatmap(data, fig_num):
    fig = px.imshow(data.corr(), labels=dict(color='Correlation'),
    x=['Movie Released Year', 'Movie Added Year', 'Duration'], y=['Movie Released Year', 'Movie Added Year', 'Duration'])
    update_fig_layout_w_legends(fig, f'{fig_num} Correlation Matrix', '', '')
    return None

def plot_heatmap_1(data, fig_num):
    fig = px.imshow(data.corr(), labels=dict(color='Correlation'),
    x=['Movie Released Year', 'Number of Directors', 'Number of Actors', 'Number of Titles'], y=['Movie Released Year', 'Number of Directors', 'Number of Actors', 'Number of Titles'])
    update_fig_layout_w_legends(fig, f'{fig_num} Correlation Matrix', '', '')
    return None

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
We have a total of 10817 unique (title, country) pairs for 124 countries. Let's plot a barplot for top 20 contributing countries. We would have to remove "Not Available" entry from the dataset to truly capture top 20's list because "Not Available" sits at position 3. Can be verified using <span style="font-family: monospace; color: #ECE93D; font-size: 12px">.index</span> attribute
</span>
<br><br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
We will also perform similar analysis for <span style="font-family: monospace; color: #ECE93D; font-size: 12px">director</span>, <span style="font-family: monospace; color: #ECE93D; font-size: 12px">cast</span>, <span style="font-family: monospace; color: #ECE93D; font-size: 12px">rating</span>, and <span style="font-family: monospace; color: #ECE93D; font-size: 12px">listed_in</span> columns and try to figure out who are busiest directors and actors and what rating and genres are popular among the audience
</span>

In [35]:
itr_list = [
    ('country', 'Not Available', 'Number of Movies + TV Shows by Country', 'Country Name', '# Movies + TV Shows', 20),
    ('director', 'Anonymous', 'Number of Movies + TV Shows by Director', 'Director Name', '# Movies + TV Shows', 20),
    ('cast', 'Anonymous', 'Number of Movies + TV Shows by Actor', 'Actor Name', '# Movies + TV Shows', 20),
    ('rating', 'Not Available', 'Number of Movies + TV Shows by Rating', 'Rating', '# Movies + TV Shows', 20),
    ('listed_in', None, 'Number of Movies + TV Shows by Genre', 'Genre', '# Movies + TV Shows', 20)
]

for idx, (col, search_phrase, title, x_title, y_title, top_n) in enumerate(itr_list):
    temp = df[['title', col]].drop_duplicates()[col].value_counts()
    x, y = np.array(temp.index), np.array(temp.values)

    if search_phrase:
        delete_at = np.where(x==search_phrase)[0][0]
        x, y = np.delete(x, delete_at), np.delete(y, delete_at)

    x, y = x[:top_n], y[:top_n]

    plot_bar_with_running_total(x, y, f'Figure {idx + 1}: ' + title, x_title, y_title)

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 1:
<br>
<br>
Inference 1: United States, India, United Kingdom, Canada, and France are the top 5 content contibuting countries.
<br>
<br>
Inference 2: More than 85% of the content comes from only 20 out of 124 countries.
<br>
<br>
Inference 3: Only 17 countries contribute 100 or more Movies / TV Shows.
<br>
<br>
<br>
<br>
Inferences From Figure 2:
<br>
<br>
Inference 4: Rajiv Chilaka has directed most number of content for Netflix and is closely followed by Jan Suter.
<br>
<br>
Inference 5: In contrast to Figure 1, we see that our data is not skewed on Directors. We see that top 20 directors out of a total of 4994 has directed only 2.5% of all the content present in our dataset.
<br>
<br>
<br>
<br>
Inferences From Figure 3:
<br>
<br>
Inference 6: Anupam Kher has acted in the most number of content for Netflix and is closely followed by Shah Rukh Khan.
<br>
<br>
Inference 7: Inline with Figure 2, we see that our data is also not skewed on Cast. We see that top 20 actors out of more than 36000 has acted only 0.8% of all the content present in our dataset.
<br>
<br>
<br>
<br>
Inferences From Figure 4:
<br>
<br>
Inference 8: TV-MA is the most preferred rating among the audience followed by TV-14.
<br>
<br>
Inference 9: Out of content available of 15 rating types, top 3 capture almost 71% of the content.
<br>
<br>
<br>
<br>
Inferences From Figure 5:
<br>
<br>
Inference 10: International Movies and International TV Shows are preferred all across the globe.
<br>
<br>
Inference 11: Drama, Comedy, and Documentary are the top 3 preferred Genres by the audience, in the order as is.
</span>

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Next, let's get some insights on <span style="font-family: monospace; color: #ECE93D; font-size: 12px">duration</span> column

In [36]:
temp_df = df.drop_duplicates(subset=['title', 'duration'])[['type', 'duration']]
movie_dist = temp_df[temp_df['type'] == 'Movie']['duration']
show_dist = temp_df[temp_df['type'] == 'TV Show']['duration']

In [37]:
plot_histogram(show_dist, 'Figure 6: Distribution of Number of Seasons', '# Seasons', '# TV Shows')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 6:
<br>
<br>
Inference 12: 2413 TV Shows out of 2672 have at a maximum of 3 Seasons. With this we could assume that audience like TV Shows with lesser seasons.
</span>

In [38]:
plot_histogram(movie_dist, 'Figure 7: Distribution of Duration (mins)', 'Duration (mins)', '# Movies')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 7:
<br>
<br>
Inference 13: We see a near perfect Normal Distribution (if we ignore observations on higher-end band of duration). The peak of the distribution is around 96 minutes with most of the movies having duration between 86 to 106 minutes.
</span>

In [39]:
temp_df = df.drop_duplicates(subset=['title', 'duration'])[['type', 'duration', 'listed_in']]
movie_dist = temp_df[temp_df['type'] == 'Movie'][['listed_in', 'duration']]
show_dist = temp_df[temp_df['type'] == 'TV Show'][['listed_in', 'duration']]

In [40]:
plot_box(show_dist['listed_in'], show_dist['duration'], 'Figure 8: Genre vs Seasons (TV Shows)', 'Genre', '# Seasons')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 8:
<br>
<br>
Inference 14: TV Shows belonging to Classic & Cult tend to be longer with half of them having atleast 4 seasons.
<br>
<br>
Inference 15: Drama, Comedy, Action & Adventure are the next three genres who tend to have longer TV Shows with half of shows of each of these categories having atleast 2 seasons.
</span>

In [41]:
plot_box(movie_dist['listed_in'], movie_dist['duration'], 'Figure 9: Genre vs Duration (Movies)', 'Genre', 'Duration (mins)')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 9:
<br>
<br>
Inference 16: Movies belonging to Romantic, LGBTQ, and Sports tend to have a very consolidated range of their duration.
<br>
<br>
Inference 17: Drama, Comedy, Action & Adventure, and Classical are the four genres who tend to have varied length of movies with duration as low as 5 minutes to as high as 250 minutes.
</span>

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Next, Let's examine the number of Movies and TV Shows added vs released each year.
</span>

In [42]:
temp_df = df.drop_duplicates(subset=['title', 'date_added', 'release_year'])[['date_added', 'release_year']]
added_year = temp_df['date_added'].apply(lambda x: x[:4]).value_counts().reset_index()
release_year = temp_df['release_year'].value_counts().reset_index()
temp_df = pd.merge(left=added_year, right=release_year, left_on='index', right_on='index', how='outer').sort_values(by='index', ascending=False).reset_index(drop=True)
temp_df = temp_df.fillna(0).rename({'index': 'Year', 'date_added': 'Added', 'release_year': 'Released'}, axis=1).astype(int)
temp_df = temp_df.melt(id_vars=['Year'], value_vars=['Added', 'Released'], var_name='Content', value_name='Count')
plot_dodge_bar(temp_df, 'Year', 'Count', 'Content', 'Figure 10: Year vs Content Released and Added', 'Year', '# Content')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 10:
<br>
<br>
Inference 18: With platform coming into existence in 2003, added its first 2 movies the very same year. Over the span of 13 years, 2003-2015, there was a constant-to-linear growth in number of content added to the platform while the growth of content released was exponential.
<br>
<br>
Inference 19: Over the next 6 years, growth of both content added and released was exponential with more content being added to the platform than released. This can be accounted for the fact that Netflix started adding Movies and TV Shows that were released in the past.
<br>
<br>
Inference 20: We saw a dip in both the number of content added and released in the recent years post 2018. While Covid can be held accountable for this decrease in the years of 2020 and 2021; we do not know why less contents were released in 2019 as compared to 2018.
</span>

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Let's examine the above graph separately for Movies and TV Shows.
</span>

In [43]:
# for movies

temp_df = df.drop_duplicates(subset=['title', 'type', 'date_added', 'release_year'])[['type', 'date_added', 'release_year']]
temp_df = temp_df[temp_df['type'] == 'Movie']
added_year = temp_df['date_added'].apply(lambda x: x[:4]).value_counts().reset_index()
release_year = temp_df['release_year'].value_counts().reset_index()
temp_df = pd.merge(left=added_year, right=release_year, left_on='index', right_on='index', how='outer').sort_values(by='index', ascending=False).reset_index(drop=True)
temp_df = temp_df.fillna(0).rename({'index': 'Year', 'date_added': 'Added', 'release_year': 'Released'}, axis=1).astype(int)
temp_df = temp_df.melt(id_vars=['Year'], value_vars=['Added', 'Released'], var_name='Movies', value_name='Count')
plot_dodge_bar(temp_df, 'Year', 'Count', 'Movies', 'Figure 11: Year vs Movies Released and Added', 'Year', '# Movies')

# for tv shows

temp_df = df.drop_duplicates(subset=['title', 'type', 'date_added', 'release_year'])[['type', 'date_added', 'release_year']]
temp_df = temp_df[temp_df['type'] == 'TV Show']
added_year = temp_df['date_added'].apply(lambda x: x[:4]).value_counts().reset_index()
release_year = temp_df['release_year'].value_counts().reset_index()
temp_df = pd.merge(left=added_year, right=release_year, left_on='index', right_on='index', how='outer').sort_values(by='index', ascending=False).reset_index(drop=True)
temp_df = temp_df.fillna(0).rename({'index': 'Year', 'date_added': 'Added', 'release_year': 'Released'}, axis=1).astype(int)
temp_df = temp_df.melt(id_vars=['Year'], value_vars=['Added', 'Released'], var_name='TV Show', value_name='Count')
plot_dodge_bar(temp_df, 'Year', 'Count', 'TV Show', 'Figure 12: Year vs TV Show Released and Added', 'Year', '# TV Show')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 11:
<br>
<br>
Inference 21: Trend is similar to what we saw in Figure 10. Similar inferences can be drawn for Movies from this figure as for Movies + TV Shows from the previous figure.
<br>
<br>
<br>
<br>
Inferences From Figure 12:
<br>
<br>
Inference 22: Trend in this figure deviate from what we saw in Figure 10 and Figure 11. Here we see a continuous rise in TV Shows released until 2020 in contrast to 2018 in the previous two figures. We can assume that this could happen either because TV Shows are longer than movies and would take more time in editting and post-shoot stuff or because TV Shows grew more popular in recent years and audience preferred them over movies.
</span>

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Next, let's plot a heatmap of correlation between content release year, content added year and, duration.
</span>

In [44]:
temp_df = df.drop_duplicates(subset=['title', 'type', 'date_added', 'release_year', 'duration'])
temp_df = temp_df[temp_df['type'] == 'Movie'][['release_year', 'date_added', 'duration']]
temp_df['date_added'] = pd.to_datetime(temp_df['date_added']).dt.year
temp_df = temp_df.astype(int)
plot_heatmap(temp_df, 'Figure 13: Movies -')

temp_df = df.drop_duplicates(subset=['title', 'type', 'date_added', 'release_year', 'duration'])
temp_df = temp_df[temp_df['type'] == 'TV Show'][['release_year', 'date_added', 'duration']]
temp_df['date_added'] = pd.to_datetime(temp_df['date_added']).dt.year
temp_df = temp_df.astype(int)
plot_heatmap(temp_df, 'Figure 14: TV Shows -')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 13:
<br>
<br>
Inference 23: Duration of movies decreased over time.
<br>
<br>
<br>
<br>
Inferences From Figure 14:
<br>
<br>
Inference 24: TV Shows with more seasons were added over time.
<br>
<br>
Inference 25: More TV Shows were released and added over time as can be verified from Figure 12.
</span>

In [45]:
temp_df = df.drop_duplicates(subset=['title', 'director', 'cast', 'release_year']).groupby(by='release_year', as_index=False).agg({'director': pd.Series.nunique, 'cast': pd.Series.nunique, 'title': pd.Series.nunique}).astype(int)
plot_heatmap_1(temp_df, 'Figure 15: ')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 15:
<br>
<br>
Inference 26: Good correlation observed observed between year and number of movies and/or tv shows released.
<br>
<br>
Inference 27: More and more directors and actors debutted over time in more and more movies and tv shows.
</span>

<span style="font-family: monospace; color: #11DDCC; font-size: 40px">
Outlier Analysis
</span>

In [46]:
m_temp = df[df['type'] == 'Movie'].drop_duplicates(subset=['title', 'duration'])[['duration']]
s_temp = df[df['type'] == 'TV Show'].drop_duplicates(subset=['title', 'duration'])[['duration']]

plot_box_one_variable(m_temp, 'Figure 16: Outlier Detection in duration (mins) of Movies','minutes')
plot_box_one_variable(s_temp, 'Figure 17: Outlier Detection in # Seasons of TV Shows','# Seasons')

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inferences From Figure 16:
<br>
<br>
Inference 28: The outlier range for movies is above 154 and below 47; and there are clearly many outliers present in the dataset having duration in these outlier range.
<br>
<br>
<br>
<br>
Inferences From Figure 17:
<br>
<br>
Inference 29: The outlier range for tv shows is above 3; and there are clearly many tv shows present in the dataset having duration in this outlier range.
<br>
<br>
</span>

<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Let's look at if any titles were added to the platform before their release date or not? ideally this should not happen but let's just check.
</span>

In [47]:
temp_df = df.drop_duplicates(subset=['title'])[['title', 'release_year', 'date_added']]
temp_df['date_added'] = pd.to_datetime(temp_df['date_added']).dt.year.astype(int)
temp_df['release_year'] = temp_df['release_year'].astype(int)
print(temp_df[temp_df['release_year'] > temp_df['date_added']].sort_values(by=['release_year', 'date_added', 'title']).reset_index(drop=True))

                         title  release_year  date_added
0                  Jack Taylor          2016        2013
1                  Tokyo Trial          2017        2016
2                       Sense8          2018        2016
3   Hans Teeuwen: Real Rancour          2018        2017
4         Arrested Development          2019        2018
5                     Incoming          2019        2018
6    Unbreakable Kimmy Schmidt          2019        2018
7              BoJack Horseman          2020        2019
8                 Fuller House          2020        2019
9           Maradona in Mexico          2020        2019
10            The Hook Up Plan          2020        2019
11                       Hilda          2021        2020
12               Love Is Blind          2021        2020
13                Polly Pocket          2021        2020


<span style="font-family: monospace; color: #77BBFF; font-size: 12px">
Inference 30: In contrast to what we believed, we saw that 14 titles were infact added on the platform before there release date. This could be a result of wrong data entry or if the trailers were made available on Netflix in name of the movie.
</span>

<span style="font-family: monospace; color: #11DDCC; font-size: 40px">
Conclusion
</span>

<hr>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 15px">
1. Defining Problem Statement and Analysing basic metrics. (10 Points)
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
Analyze the data and generate insights that could help Netflix in deciding which type of TV Show/Movie to produce and how to grow the business in different countries.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
Method 1: Exploratory Data Analysis - Knowing the data; filling in missing data; transforming the data; 
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
Method 2: Uni-Variate Analysis - Histograms; Count Plots
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
Method 3: Multi-Variate Analysis - Bar Plots; Line Plots; Box Plots; Pair Plots
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
Method 4: Correlation - Heatmaps; Correlation Matrix
</span>
<br>
<br>
<hr>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 15px">
2: Observations on the shape of data, data types of all the attributes, conversion of categorical attributes to 'category' (If required), missing value detection, statistical summary. (10 Points)
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 14px; padding-left: 3em">
Statistical Summary of the Raw Data:
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
type - 2 unique values - 'Movie' and 'TV Show' with frequency of 6131 and 2676 respectively.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
title - 8807 unique values.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
director - 4528 unique values with 'Rajiv Chilaka' occurring 19 times. 2634 NULL values.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
cast - 7692 unique values with David Attenborough occurring 19 times. 825 NULL values.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
country - 748 unique values with 'United States' occurring 2818 times. 831 NULL values.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
date_added - 1714 unique values with '2020-01-01' occurring 110 times. 10 NULL values.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
release_year - 74 unique values with '2018' occurring 1147 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
rating - 17 unique values with 'TV-MA' occurring 3207 times. 4 NULL values.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
duration - 220 unique values with '1 Season' occurring 1793 times. 3 NULL values.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
listed_in - 514 unique values with 'Dramas, International Movies' occurring 362 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
description - 8775 unique values with 'Paranormal activity at a lush, abandoned prope...' occurring 4 times.
</span>
<br>
<br>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 14px; padding-left: 3em">
Statistical Summary of the Processed Data:
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
type - 2 unique values - 'Movie' and 'TV Show' with frequency of 145305 and 56011 respectively.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
title - 8777 unique values with 'Kahlil Gibran's The Prophet' occurring 700 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
director - 4994 unique values with 'Anonymous' occurring 50497 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
cast - 36433 unique values with 'Anonymous' occurring 2139 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
country - 124 unique values with 'United States' occurring 59214 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
date_added - 1719 unique values with '2020-01-01' occurring 3748 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
release_year - 74 unique values with '2018' occurring 24198 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
rating - 15 unique values with 'TV-MA' occurring 73675 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
duration - Max value = 312; Min Value = 1; Mean Value = 77.58; Median Value = 95
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
listed_in - 42 unique values with 'Dramas' occurring 29713 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
description - 8775 unique values with 'A troubled young girl and her mother find sola...' occurring 700 times.
</span>
<br>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px;">

| Property of the Data | Raw Data Statistics | Processed Data Statistics |
| -------------------- | ------------------- | ------------------------- |
| Shape of the Data | (8807, 12) | (201364, 11) |
| Data types of columns | int - release_year<br> object - all other columns | int - duration<br> object - all other columns |
| Conversion of data types | release_year: int<br> duration: object | release_year: object<br> duration: int |
| Missing Values | Present in director, cast, country, date_added, rating, duration | None Present|

</span>
<br>
<hr>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 15px">
3. Non-Graphical Analysis: Value counts and unique attributes. ​​(10 Points)
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
a. There are almost 2.5 times Movie titles in our dataset than TV shows.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
b. On exploding our data, a row in the original dataframe expanded to 700 for title Kahlil Gibran's The Prophet while few remained they were
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
c. Martin Scorsese occurs the most - 419 - number of times in our dataset; closely followed by Youssef Chahine - 409 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
d. Liam Neeson occurs the most - 161 - number of times in our dataset; closely followed by Alfred Molina - 160 times.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
e. Highest number of entries - 59214 - in our dataset corresponds to the US followed by India at 22733.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
f. Most of the movies are added on the platform on 1st of the months.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
g. More movies from recent years are present in our dataset.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
h. Highest number of entries - 73695 - in our dataset corresponds to the TV-MA followed by TV-14 at 43597.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
i. Highest number of entries - 29713 - in our dataset corresponds to the Dramas followed by International Movies at 28084.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
j. Note: Number of unique values in all these columns are mentioned in the Statistical Summary in the asnwer of Q2.
</span>
<br>
<br>
<hr>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 15px">
4. Visual Analysis - Univariate, Bivariate after pre-processing of the data. Note: Pre-processing involves unnesting of the data in columns like Actor, Director, Country.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
Refer to Figure 1 through 15.
</span>
<br>
<br>
<hr>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 15px">
5. Missing Value & Outlier check (Treatment optional). (10 Points)
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
Missing values were observed in columns director, cast, country, date_added, rating, duration columns and were treated as follows:
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
-  For columns director and cast, NaNs were replaced with "Anonymous"
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- For columns country and rating, NaNs were replaced with "Not Available"
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- For column date_added, NaTs were replaced with 31st December of the content released year.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- For column duration, NaNs were a result of wrong data entry; these values were present in the corresponding rows of rating column.
</span>
<br>
<br>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
Outliers were observed in columns duration and date_added & release_year.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- Refer to Figure 16 and 17 and Inference 28 to 30.
</span>
<br>
<br>
<hr>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 15px">
6. Insights based on Non-Graphical and Visual Analysis. (10 Points)
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
Refer to Inference A to I and 1 to 30.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
Comments on range of columns - date_added, release_year, and duration.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- Netflix added its first content on Dec 31st, 2003. The latest content was uploaded on the platform on Sept 25th, 2021
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- Netflix added its first content on Dec 31st, 2003. The latest content was uploaded on the platform on Sept 25th, 2021
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- Duration of a movie on Netflix ranges from 3 to 312 minutes.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- TV Shows on Netflix have a mximum of 17 seasons while minimum being 1.
</span>
<br>
<br>
<hr>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 15px">
7. Business Insights (10 Points) - Should include patterns observed in the data along with what you can infer from it.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">
Please refer to Infrence 1 to 30 above for detailed answer of this question. Some of the most valuable insights are written below.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- International Movies and TV Shows are preferred by audience all across the globe.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- TV-MA, TV-14, and R are the most common ratings among the customer-base of Netflix.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- Movies and TV Shows from US, India, and UK are the most popular ones among the people.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- Dramas, Comedies, and Documentaries are the most common genres for TV Shows.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- Dramas, Comedies, Documentaries, and Classical are the most common genres for Movies.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 5em">
- Audience love watching TV Shows with less then 2 seasons and Movies about 100 minutes long.
</span>
<br>
<br>
<hr>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 15px">
8. Recommendations (10 Points) - Actionable items for business. No technical jargon. No complications. Simple action items that everyone can understand.
</span>
<br>
<br>
<span style="font-family: monospace; color: #77BBFF; font-size: 12px; padding-left: 3em">

- Netflix can introduce regional content in most popular genres (Dramas, Comedies, and Documentaries) and ratings (TV-MA, TV-14, and R) to increase its customer base in less popular regions for example Europe and Africa.

- Netflix can produce longer movies and shows in the popular genres - Dramas, Comedies, and Documentaries - because these genres are loved by the audience and new shows will not find it hard to find people's love. This would account for recurring payments for Netflix because longer shows will span over months or years and people would have to pay to watch what they love.

- Since most of the movies and shows on Netflix are added in the first week of the month, the Engineering Team should maintain and scale servers accordingly for this week and also for the rest of the month.

- Since the database does not have many old movies or shows, it is worth the shot to add these contents on Netflix to see if they get any traction from the audience.

- Add new movies and shows belonging to less popular ratings and genres to give audience more options in these unppopular areas to watch from and who know they may love the new content.

- Netflix can introduce a weekly subscription plan that is valid only for the first week of a month. Since most of the new movies and shows are added in the first week, people signing up for this subscription will try to consume as much content as possible and this could form a habit and thus produce a loyal customer of Netflix.

- Netflix can bring together most popular cast members and directors under one umbrella to produce shows and movies that would be popular and accepted by a very large base of audience.
</span>
<br>

<hr>